api/incubator/src/main/java/io/opentelemetry/api/incubator/config/ConfigProvider.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.api.incubator.config;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * A registry for accessing configuration.
+ *
+ * <p>The name <i>Provider</i> is for consistency with other languages and it is <b>NOT</b> loaded
+ * using reflection.
+ *
+ * <p>See {@link InstrumentationConfigUtil} for convenience methods for extracting config from
+ * {@link ConfigProvider}.
+ */
+@ThreadSafe
+public interface ConfigProvider {
+
+  /**
+   * Returns the {@link StructuredConfigProperties} corresponding to <a
```

Comment:
I don't know enough about the domain yet. There are definitely use cases for dynamic configuration, where an SDK would connect to a remote config server via [opamp](https://github.com/open-telemetry/opamp-spec), and periodically receiver commands to update the config. But a lot needs to be built to enable that, including an opamp client implementation in java, and SDK APIs for updating SdkMeterProvider, SdkLoggerProvider, SdkTracerProvder which are currently all statically configured.

Then there's the question of what types of dynamic config scenarios even make sense for instrumentation. If the otel java agent makes decisions to install or omit bytecode instrumentation based on ConfigProvider, it won't always be possible to go install later if the config were to change. Other instrumentation options may be better fits for dynamic config, but then there's the question of how often to check for updates since we probably don't want to be looking for / applying config updates on the hot path.

Reply:
not to anchor things too much, but certain dynamic elements may be better feature driven anyway. For example, recently in the zipkin java stuff there were requests for the sender (exporter here) to have dynamic endpoints (supplier function), provided by eureka etc. So, this is a code change as basically deferring endpoint resolution isn't default in anything. After that, there's what is the initial value as well. Point being is indeed components need to change to be dynamic, but specifically what they are dynamic about may help flex the api. Meanwhile basic impl that's easy to reason with could be a step forward.